activerecord/lib/arel/table.rb

```diff
@@ -10,11 +10,15 @@ class << self; attr_accessor :engine; end
 
     attr_accessor :name, :table_alias
 
-    # TableAlias and Table both have a #table_name which is the name of the underlying table
-    alias :table_name :name
-
     def initialize(name, as: nil, klass: nil, type_caster: klass&.type_caster)
-      @name = name.to_s
+      @name =
+        case name
+        when Symbol then name.to_s
+        when Nodes::Node then Arel.sql(name.to_sql)
```

Comment:
I decided to not carry an ast node around and convert it into a SQL literal right here.

Alternatively, I considered updated the SQL visitor (`to_sql.rb`) and handle nodes in the `#quote_table_name` method. But that would affect some other nodes, too, so I hesitated.

Reply:
The downside of calling `.to_sql` and wrapping it in a `Arel::Nodes::SqlLiteral` is that this looses track of potential bind params inside the expression. This might be undesirable, especially because the examples you bring up are function calls with parameters. 

By not being able to use bind params, the resulting query will get a different signature for every set of params and can't be as easily reused as a prepared statement.

That being said, I agree that `#quote_table_name` probably isn't the best place to implement this.

Reply:
> the resulting query will get a different signature for every set of params and can't be as easily reused as a prepared statement.

That's a good point (though in real life I prefer to avoid prepared statements). 

Refactored to perform complication within a visitor (`to_sql.rb#visit_Arel_Table`).